include/openthread/tcp.h

```diff
@@ -0,0 +1,521 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread TCP API.
+ *
+ */
+
+#ifndef OPENTHREAD_TCP_H_
+#define OPENTHREAD_TCP_H_
+
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-tcp
+ *
+ * @brief
+ *   This module includes functions that control TCP communication.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This a vectored packet structure for use with TCP.
+ *
+ * A single @p otLinkedBuffer structure references an array of bytes in memory,
+ * via @p mData and @p mLength. The @p mNext field is used to form a chain of
+ * @p otLinkedBuffer structures.
+ */
+typedef struct otLinkedBuffer
+{
+    uint8_t* mData;              ///< Pointer to data referenced by this linked buffer
```

Comment:
How about defining it as `const uint8_t *mData`?

In both tx and rx uses of `otLinkBuffer` after an instance of `otLinkBuffer` is created, and as we pass it to different APIs, I don't think we want to allow the `mData` content to be changed.

Reply:
I'm OK with this. Any thoughts about whether making this `const` might make it harder for the application developer, since they can't modify data via this pointer? For example, if the application wants to modify data in-place in the receive buffer while processing it, maybe we shouldn't disallow it. Similarly, the application developer won't be able to use the `mData` pointer when initializing the buffer with data to send. I think there are strong arguments either way.

If there's broad agreement that we should make this `const` then I'm happy to include this change. Any thoughts?

Reply:
The owner/creator of a `otLinkedBuffer` instance should already be able to write the data in the buffer it allocates/gets (and sets a mData`, but I think once we pass the `otLinkedBuffer` instance to some other entity we don't want to let it to be able to change the content anymore. In general in OT, we aim to follow `const` correctness. This helps avoid incorrect use and potential bugs.

When sending, we can actually have `mData` point to some constant/fixed array in flash.

Reply:
I agree that is a valid reason. I think you've convinced me that using `const` is a good idea. :)

I'll push an update to this pull request changing `mData` to const as you suggested.

Reply:
(To clarify, this does restrict users from modifying the receive buffer contents in place when processing it. But we can always remove a `const` later. Adding `const` at a later time may break applications, so it seems better to include it now.)

Reply:
@abtink Are the updated documentation comments OK?

Reply:
This was pushed in an update yesterday. I'm leaving this conversation open for a while in case someone wants to object to this change, but if nobody objects I'll resolve this conversation.

Reply:
It seems nobody has objected, so I'm going to resolve this conversation now.